frontend/controller/serviceworkers/push.js

```diff
@@ -0,0 +1,136 @@
+      const pubsub = sbp('okTurtles.data/get', PUBSUB_INSTANCE)
+      if (!pubsub) throw new Error('Missing pubsub instance')
+
+      const readyState = pubsub.socket.readyState
+      if (readyState !== WebSocket.OPEN) {
+        throw new Error('WebSocket connection is not open')
+      }
+
+      const socket = pubsub.socket
+      const reported = map.get(socket)
+      map.set(socket, subscriptionInfo)
+      if (subscriptionInfo?.endpoint && reported !== subscriptionInfo.endpoint) {
+        // If the subscription has changed, report it to the server
+        pubsub.socket.send(createMessage(
+          REQUEST_TYPE.PUSH_ACTION,
+          { action: PUSH_SERVER_ACTION_TYPE.STORE_SUBSCRIPTION, payload: subscriptionInfo }
+        ))
+      } else if (!subscriptionInfo && reported) {
+        // If the subscription has been removed, also report it to the server
+        pubsub.socket.send(createMessage(
+          REQUEST_TYPE.PUSH_ACTION,
+          { action: PUSH_SERVER_ACTION_TYPE.DELETE_SUBSCRIPTION, payload: null }
+        ))
+      }
+    }
+  })()
+}): string[])
+
+if (self.registration?.pushManager) {
+  (() => {
+    let inProgress = false
+    sbp('okTurtles.events/on', PUBSUB_RECONNECTION_SUCCEEDED, () => {
+      if (inProgress) return
+      inProgress = true
+      self.registration.pushManager.getSubscription().then((subscription) =>
+        sbp('push/reportExistingSubscription', subscription?.toJSON())
+      ).catch((e) => {
+        console.error('Error reporting subscription on reconnection', e)
+      }).finally(() => {
+        inProgress = false
+      })
+    })
+  })()
+}
+
+self.addEventListener('push', function (event) {
+  // PushEvent reference: https://developer.mozilla.org/en-US/docs/Web/API/PushEvent
+  if (!event.data) return
+  const data = event.data.json()
+  if (data.type === NOTIFICATION_TYPE.ENTRY) {
+    event.waitUntil(setupChelonia().then(() => {
+      // We have event data, so we process it
+      if (data.data) return sbp('chelonia/handleEvent', data.data)
+      // We just sync the contract if there's no event data. Sync could fail if
+      // there are no references, hence the catch
+      return sbp('chelonia/contract/sync', data.contractID).catch(e => {
+        console.error('[push event] Error syncing', data.contractID, e)
+      })
+    }).catch((e) => {
+      console.error('Error processing push event', e)
+      if (data.contractType === 'gi.contracts/chatroom') {
+        return makeNotification({ title: '@@@err', body: e.message })
+      }
+    }))
+  }
+}, false)
+
+self.addEventListener('pushsubscriptionchange', async function (event) {
+  // NOTE: Currently there is no specific way to validate if a push-subscription is valid. So it has to be handled in the front-end.
+  // (reference:https://pushpad.xyz/blog/web-push-how-to-check-if-a-push-endpoint-is-still-valid)
+  const subscription = await self.registration.pushManager.subscribe(
+    event.oldSubscription.options
+  )
```

Comment:
Should there be a `.catch` here or something?

Reply:
There is now a catch at the end